Error handling

peppsac opened the pull request:
This PR implements error handling for layers loading.

It introduces a new `LayerUpdateState` object holding fetch info: error count, last error timestamp etc.
`NodeProcess` can them use this to decide if it can tries to download texture from a given layer.

Currently there's a single error strategy implemented: it simply retries after longer and longer wait period. We could later implement different strategies if needed.

peppsac committed:
feat: Throw custom error when cancelling commands

So the caller can act differently if needed.

peppsac committed:
refactor: Invoke command.resolve/reject in ManagerCommand

Doing this in a single place makes sure that all Providers results
are treated the same way.

peppsac committed:
feat: Add LayerUpdateState object to manager updates cycle

Maintaining a state (success, pending, error) to make better informed
decision: should I update this layer? did it failed earlier? etc

peppsac committed:
refactor: Stop catching errors in WMS/WTMS providers

Both providers don't have enough information to properly handle errors
so stop catching them.

peppsac committed:
refactor: Use new LayerUpdateState

Improve error management.

peppsac committed:
refactor: Elevation textures can now have 3 states

> undefined
    initial state
> emptyTexture (level == -1)
    Texture was updated but we couldn't find a good texture. This is a
    final state and we can't improve it. (FIXME)
> texture (level > 0)
    There's a texture

peppsac committed:
feature: Add 'needsRedraw' param to notifyChange

This way caller can indicate if the change needs a scene redraw, or simply
a scene update.
Example usage: if a command fails, we want to schedule a new update, but we
don't need a redraw (since nothing changed)

nosy-b commented:
When zooming in the app with default layers it raises:
 `Uncaught (in promise) TypeError: Cannot read property 'scene' of undefined
    at NodeProcess.js:413`
Three is still r80, is that normal?

nosy-b commented:
- Camera rotation (ctrl-mouseMove) is not working as before. Sometimes it doesn't take the projected center of the screen as pivot.

peppsac commented:
I fixed the `Uncaught (in promise) TypeError: Cannot read property 'scene' of undefined at NodeProcess.js:413` problem.

nosy-b commented:
This PR is a good idea because it is necessary to efficiently handle requests when no data is available but:
- In _LayerUpdateState.js_ 
  `const PAUSE_BETWEEN_ERRORS = [ 1000.0, 3000.0, 7000.0, 6000.0 ];`
  is not longer and longer ;)
- The _PAUSE_BETWEEN_ERRORS_  has no impact. It will still infinitely loop when you go outside of available data. The only way to kill this is to return false on update_error but it also block the nodeprocess.

The test I usually use: 
Zoom strongly to italy -> it launches infinite tile requests 
Zoom to french  DOM-TOM where high resolution data is available for the default index services->too much tiles in command to process actual visible ones

peppsac commented:
> In LayerUpdateState.js const PAUSE_BETWEEN_ERRORS = [ 1000.0, 3000.0, 7000.0, 6000.0 ]; is not longer and longer ;)

The last value was supposed to be `60000`...
And yes it's not strictly `longer and longer` - but a 60 sec pause beween retries seems like a good first step.

> The PAUSE_BETWEEN_ERRORS has no impact. It will still infinitely loop when you go outside of available data. The only way to kill this is to return false on update_error but it also block the nodeprocess.

Do you mean that it doesn't wait before retrying a request?

I'll try your testing strategy (thanks for sharing it) :)

nosy-b commented:
> Do you mean that it doesn't wait before retrying a request?

Yes it will request the same tiles (unavailable) infinitely at high speed

We should make the master version with a default behaviour that simply cut the request on _update_error_ or a too long _pending loading_ so _nodeprocess_ don't retry any command and don't go down the tree of this tile.

peppsac commented:
I pushed a fix for 2 typos in the code, it should work better now :)

peppsac commented:
Keep in mind that it's a first step toward error handling. The goal is to take a first stab at it and agree on a plan.

Also I know @autra is working on a PR unifying network accesses: it'll help with error handling.

nosy-b commented:
Ok I'll check future @autra PR.
For your last fix, it does stop the unlimited request when no data is available that's a good point!
But if you pause action with no data and zoom in again it will try new requests where ideally it should not go down the tree if the leaf has no data available. (Also still always impossible to see Ile de La Réunion also for example).
For the camera rotation bug do you know where it comes from, maybe not this PR? It happens when zooming from far and then rotating the camera, sometimes the rotation is not centered and if you do it again without moving it'll be centered.

autra commented:
> For the camera rotation bug do you know where it comes from, maybe not this PR? It happens when zooming from far and then rotating the camera, sometimes the rotation is not centered and if you do it again without moving it'll be centered.

FWIW I have this bug on master too, I seem to always go back to Africa ;-)

gchoqueux commented on src/Core/Commander/ManagerCommands.js:
```diff
@@ -178,7 +180,13 @@ ManagerCommands.prototype.getProviders = function() {
 
+CancelledCommandException.prototype.toString = function() {
+    return 'Cancelled command ' + this.command + ' for node ' + this.command.requester;
+}
```
add semi colon

gchoqueux commented on src/Core/Commander/ManagerCommands.js:
```diff
@@ -178,7 +180,13 @@ ManagerCommands.prototype.getProviders = function() {
     return p;
 };
 
+function CancelledCommandException(command) {
```
Can you explain the addition of this object?

gchoqueux commented on src/Core/Commander/Providers/WMTS_Provider.js:
```diff
@@ -133,20 +133,17 @@ WMTS_Provider.prototype.getXbilTexture = function(tile, layer, parameters) {
-        return Promise.resolve(null);
+        let minmax = this._IoDriver.computeMinMaxElevation(
+            textureCache.floatArray,
+            256, 256,
```
Create constant like :
```js
const SIZE_TEXTURE_TILE = 256;
```

gchoqueux commented on src/Renderer/LayeredMaterial.js:
```diff
@@ -320,9 +320,9 @@ LayeredMaterial.prototype.setNbLayersColor = function(n) {
 
 
 LayeredMaterial.prototype.getLayerLevel = function(id, slot) {
-    var level = this.Textures[id][slot || 0].level;
```
var or let?

gchoqueux commented on src/Scene/BrowseTree.js:
```diff
@@ -119,10 +119,13 @@ BrowseTree.prototype._browseNonDisplayableNode = function(node, level, process,
+    if (!node.loaded) {
         // Make sure this node is not stuck in a !loaded state
+        // TODO: we could be smarter and do this only for visible tiles for instance
+        // but we need to be careful due to the number of possible states for a tile (pendingSubdivision,
```
:+1:

gchoqueux commented on src/Scene/LayerUpdateState.js:
```diff
@@ -0,0 +1,57 @@
+const UPDATE_IDLE = 0;
```
little comment for this new object

autra replied:
Also, you can consider doing an "enum" (well, js-style):

```js
const UPDATE_STATE = {
    IDLE: 0,
    PENDING: 1,
    // ...
};
```

gchoqueux commented on src/Scene/LayerUpdateState.js:
```diff
@@ -0,0 +1,57 @@
+            return errorDuration <= (timestamp - this.lastErrorTimestamp);
+        }
+    }
+}
```
Missing semicolon

gchoqueux commented on src/Scene/LayerUpdateState.js:
```diff
@@ -0,0 +1,57 @@
+        Math.max(0, Math.min(this.errorCount, PAUSE_BETWEEN_ERRORS.length) - 1);
+
+    return PAUSE_BETWEEN_ERRORS[idx];
+}
```
Missing semicolon

gchoqueux commented on src/Scene/LayerUpdateState.js:
```diff
@@ -0,0 +1,57 @@
+
+LayerUpdateState.prototype.try = function() {
+    this.state = UPDATE_PENDING;
+}
```
Missing semicolon

gchoqueux commented on src/Scene/LayerUpdateState.js:
```diff
@@ -0,0 +1,57 @@
+LayerUpdateState.prototype.success = function() {
+    this.lastErrorTimestamp = 0;
+    this.state = UPDATE_IDLE;
+}
```
Missing semicolon

gchoqueux commented on src/Scene/LayerUpdateState.js:
```diff
@@ -0,0 +1,57 @@
+    this.lastErrorTimestamp = timestamp;
+    this.state = UPDATE_ERROR;
+    this.errorCount++;
+}
```
Missing semicolon

gchoqueux commented on src/Scene/NodeProcess.js:
```diff
@@ -290,23 +274,33 @@ function updateNodeImagery(quadtree, node, layersConfig, force) {
+                    node.layerUpdateState[layer.id].success();
+                } else {
+                    node.layerUpdateState[layer.id].failure(Date.now());
+                    scene.notifyChange(node.layerUpdateState[layer.id].waitDurationUntilNextTry(), false);
```
We could instead dispatch an event system to notify a change.

peppsac replied:
Changing from `notifyChange` function to an event belongs to another PR (if we decide to do the change).

gchoqueux commented:
This PR is in progress? it needs to rebase?

peppsac commented:
I fixed all your comments on a local branch except the last one (using an event).
And of course this PR needs a rebase...

(I removed the `ready` flag)

gchoqueux commented:
You will push your changes?

autra commented on src/Core/Commander/Providers/WMS_Provider.js:
```diff
@@ -145,12 +142,12 @@ WMS_Provider.prototype.getXbilTexture = function(tile, layer, bbox, pitch) {
 
     if (textureCache !== undefined) {
-        return Promise.resolve(textureCache ? {
+        return Promise.resolve({
```
In this case, I would replace the if by 

```js
if (textureCache) {
```
Covering also the null case (and false and empty string). Btw the old code was resolving with null when textureCache was null. Was it used somewhere?

peppsac replied:
`null` value for elevation textures is gone with commit `Elevation textures can now have 3 states`.
 It was a special value used for error handling (sort-of).

autra commented on src/Core/Commander/Providers/WMTS_Provider.js:
```diff
@@ -133,20 +133,17 @@ WMTS_Provider.prototype.getXbilTexture = function(tile, layer, parameters) {
     var textureCache = this.cache.getRessource(url);
 
     if (textureCache !== undefined) {
```
`if (textureCache)` ?

autra commented on src/Scene/LayerUpdateState.js:
```diff
@@ -0,0 +1,57 @@
+
+}
+
+LayerUpdateState.prototype.constructor = LayerUpdateState;
```
Only needed if you assign something to the prototype, which is not the case here.

autra commented on src/Scene/LayerUpdateState.js:
```diff
@@ -0,0 +1,57 @@
+}
+
+LayerUpdateState.prototype.waitDurationUntilNextTry = function() {
+    if (this.state != UPDATE_ERROR) {
```
It's a good practice to use `!==` and `===` by default. `==` *can* be useful, but js implicit conversion rules are kind of... surprising sometimes, so if we can avoid them, it's better.

autra commented on src/Scene/NodeProcess.js:
```diff
@@ -432,7 +456,7 @@ NodeProcess.prototype.processNode = function(node, camera, params) {
 
     return wasVisible || isVisible;
-}
+};
```
<3

autra commented on src/Scene/LayerUpdateState.js:
```diff
@@ -0,0 +1,57 @@
+    }
+}
+
+LayerUpdateState.prototype.waitDurationUntilNextTry = function() {
```
Not very fan of this method name. You're returning a time right? You're not actually waiting. `getWaitTimeUntilNextTry`?

autra replied:
Ok actually `wait` was a noun :-) I think starting by `get` is a bit clearer.

peppsac replied:
I don't use `get`  for method that aren't accessors.
But I agree with your comment so I changed the name to: `secondsUntilNextTry()`

autra commented on src/Scene/LayerUpdateState.js:
```diff
@@ -0,0 +1,57 @@
+        return 0;
+    }
+    let idx =
+        Math.max(0, Math.min(this.errorCount, PAUSE_BETWEEN_ERRORS.length) - 1);
```
If I'm understanding correctly, you never give up, but always try at least once per min right?

peppsac replied:
Yes

peppsac commented:
Long overdue branch update done :-) !

gchoqueux commented on src/Scene/NodeProcess.js:
```diff
@@ -280,22 +289,32 @@ function updateNodeImagery(quadtree, node, layersConfig, force) {
+                    node.layerUpdateState[layer.id].success();
+                } else {
+                    node.layerUpdateState[layer.id].failure(Date.now());
+                    scene.notifyChange(node.layerUpdateState[layer.id].secondsUntilNextTry() * 1000, false);
```
Can you comment for next PR?
to replace `scene.notifyChange` by event.

```js
foo.dispatchEvent(eventLayerUpdateFailure);
```

gchoqueux commented on src/Scene/Node.js:
```diff
@@ -21,7 +21,7 @@ function Node() {
     this.screenSpaceError = 0.0;
     this.loaded = false;
     this.pendingSubdivision = false;
```
it's possible to remove `this.pendingSubdivision` and use `this.layerUpdateState = {};`?

peppsac replied:
Yes, but it requires quite a lot of refactoring because `pendingSubdivision` is used in several places that doesn't know about layers (eg TileMesh, Node), while `layerUpdateState` needs a layer.

I'll probably do it later though.

gchoqueux replied:
add TODO in the code and new issue

gchoqueux commented on src/Core/Commander/ManagerCommands.js:
```diff
@@ -84,7 +89,7 @@ ManagerCommands.prototype.runCommand = function runCommand(command, queue, execu
         // to reduce CPU load (no need to perform an update on completion if we
         // know there's another one ending soon)
-        this.scene.notifyChange(100);
+        this.scene.notifyChange(100, true);
```
just for info : hard code value
needs TODO comment?

peppsac replied:
There's a 3 line comment above this value, explaining the rationale.

gchoqueux replied:
`this.scene.notifyChange(100);` 
use `export const delayUpdate = 100`
and This value (100) may change depending on the computer

peppsac replied:
This change comes from an earlier commit (4218023e).

gchoqueux commented on src/Renderer/LayeredMaterial.js:
```diff
@@ -369,12 +369,20 @@ LayeredMaterial.prototype.isColorLayerDownscaled = function isColorLayerDownscal
+            return true;
+        }
+        //   - blank texture (eg: empty xbil texture)
+        if (tex.level === -1) {
```
just for info : hard code value
needs TODO comment?

gchoqueux commented on src/Scene/LayerUpdateState.js:
```diff
@@ -0,0 +1,60 @@
+ * It stores information to allow smart update decisions, and especially network
+ * error handling.
+ */
+function LayerUpdateState() {
```
You can try ES6 Class, if you want

gchoqueux commented on src/Scene/NodeProcess.js:
```diff
@@ -12,12 +12,13 @@ import Projection from 'Core/Geographic/Projection';
-    // Constructor
-
+function NodeProcess(scene, camera, ellipsoid, bbox) {
+    this.scene = scene;
```
TODO in new PR, remove `this.scene`
